File: leetcode/medium/229.majority-element-ii.py
#
# @lc app=leetcode id=229 lang=python3
#
# [229] Majority Element II
#

# @lc code=start
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        count_one = 0
        count_two = 0
        ans=[0]*2
        result=[]
        n = len(nums)
#finding probable answer values
        for i in range(0, n):
            if (count_one==0 and ans[1]!=nums[i]):
                ans[0]=nums[i]
                count_one=1
            elif (count_two==0 and ans[0]!=nums[i]):
                ans[1]=nums[i]
                count_two=1
            elif (ans[0]==nums[i]):
                count_one+=1
            elif  (ans[1]==nums[i]):
                count_two+=1
            else:
                count_one-=1
                count_two-=1

        
        count_a=0
        count_b=0

        #verification

        for i in range(0, len(nums)):
            if (ans[0]==nums[i]):
                count_a+=1
            elif(ans[1]==nums[i]):
                count_b+=1
            
        #edge cases, sometimes, we dont get two values > n/3 or cases like [1,2,3]
        if (count_a>n/3):
            result.append(ans[0])
        if(count_b>n/3):
            result.append(ans[1])

            
        

        return result



        
# @lc code=end

File: leetcode/229.majority-element-ii.cpp
/*
 * @lc app=leetcode id=229 lang=cpp
 *
 * [229] Majority Element II
 */

// @lc code=start
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        int count_one=0;
        int count_two=0;
        int n = nums.size();
        int candidate1=-1;
        int candidate2=-1;
        // std::vector<int> ans(2);
        std::vector<int> result;

        for (int i=0; i<n; i++) {
            if (count_one==0 && candidate2!=nums[i]) {
                candidate1=nums[i];
                count_one=1;
            } else if (count_two==0 && candidate1!=nums[i]) {
                candidate2=nums[i];
                count_two=1;
            } else if (candidate1==nums[i]) {
                count_one++;
            } else if (candidate2==nums[i]) {
                count_two++;
            } else {
                count_one--;
                count_two--;
            }
        }

        int count_a=0;
        int count_b=0;

        for (int i = 0; i<n; i++) {
            if (candidate1==nums[i]) {
                count_a++;
            } else if (candidate2==nums[i]) {
                count_b++;
            }
        }

        if (count_a>n/3) {
            result.push_back(candidate1);
        }
        if (count_b>n/3) {
            result.push_back(candidate2);
        }

        return result;

        
        
    }
};
// @lc code=end

